### ISSUE TITLE
Crash when using CacheDataSource

### ISSUE THREAD

**Issue by Hiwensen** [google/ExoPlayer#6588]
### [REQUIRED] Issue description
Thanks for developing such a great app, I learned a lot from it.
But now I suffered a crash with it after using the `CacheDataSource` and related class

This is the code I use for adding cache for player:
```
fun buildCacheDataSourceFactory(context: Context, bandwidthMeter: DefaultBandwidthMeter?): DataSource.Factory {
            val httpDataSourceFactory =
                    DefaultHttpDataSourceFactory(Util.getUserAgent(context, APPLICATION_NAME), bandwidthMeter)
            val cache = getCacheSingleInstance(context)
            return CacheDataSourceFactory(cache, httpDataSourceFactory, CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR)
        }
        
        private fun getCacheSingleInstance(context: Context): Cache? {
            var cache: Cache? = null
            val cachePath = context.cacheDir.absolutePath + File.separator + PLAYER_CACHE_SUBDIRECTORY
            val cacheFile = File(cachePath)
            val isLocked = SimpleCache.isCacheFolderLocked(cacheFile)
            if (!isLocked) {
                cache =
                        SimpleCache(cacheFile, LeastRecentlyUsedCacheEvictor(DEFAULT_MAX_CACHE_SIZE), ExoDatabaseProvider(context))
            }
            return cache
        }
```

### [REQUIRED] Reproduction steps
This is an occasional crash. I haven't reproduced this crash yet, but from the crashlytics of the fabric, more than 600 users suffered it with more than 1800 times in the past 7 days.

### [REQUIRED] A full bug report captured from the device
Here below are the detailed crash message:
Crash 1, happened on multiple devices with system version 4, 5, and 6.
```
Fatal Exception: java.lang.StackOverflowError
       at java.util.WeakHashMap$Entry.<init> + 71(WeakHashMap.java:71)
       at java.util.WeakHashMap.put + 611(WeakHashMap.java:611)
       at android.database.sqlite.SQLiteConnectionPool.finishAcquireConnectionLocked + 911(SQLiteConnectionPool.java:911)
       at android.database.sqlite.SQLiteConnectionPool.tryAcquirePrimaryConnectionLocked + 847(SQLiteConnectionPool.java:847)
       at android.database.sqlite.SQLiteConnectionPool.waitForConnection + 613(SQLiteConnectionPool.java:613)
       at android.database.sqlite.SQLiteConnectionPool.acquireConnection + 348(SQLiteConnectionPool.java:348)
       at android.database.sqlite.SQLiteSession.acquireConnection + 894(SQLiteSession.java:894)
       at android.database.sqlite.SQLiteSession.prepare + 586(SQLiteSession.java:586)
       at android.database.sqlite.SQLiteProgram.<init> + 58(SQLiteProgram.java:58)
       at android.database.sqlite.SQLiteStatement.<init> + 31(SQLiteStatement.java:31)
       at android.database.sqlite.SQLiteDatabase.insertWithOnConflict + 1467(SQLiteDatabase.java:1467)
       at android.database.sqlite.SQLiteDatabase.replaceOrThrow + 1411(SQLiteDatabase.java:1411)
       at com.google.android.exoplayer2.upstream.cache.CacheFileMetadataIndex.set + 166(CacheFileMetadataIndex.java:166)
       at com.google.android.exoplayer2.upstream.cache.SimpleCache.startReadWriteNonBlocking + 422(SimpleCache.java:422)
       at com.google.android.exoplayer2.upstream.cache.SimpleCache.startReadWriteNonBlocking + 49(SimpleCache.java:49)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource + 384(CacheDataSource.java:384)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 317(CacheDataSource.java:317)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)

Crashed: Loader:HlsSampleStreamWrapper
       at java.util.WeakHashMap$Entry.<init> + 71(WeakHashMap.java:71)
       at java.util.WeakHashMap.put + 611(WeakHashMap.java:611)
       at android.database.sqlite.SQLiteConnectionPool.finishAcquireConnectionLocked + 911(SQLiteConnectionPool.java:911)
       at android.database.sqlite.SQLiteConnectionPool.tryAcquirePrimaryConnectionLocked + 847(SQLiteConnectionPool.java:847)
       at android.database.sqlite.SQLiteConnectionPool.waitForConnection + 613(SQLiteConnectionPool.java:613)
       at android.database.sqlite.SQLiteConnectionPool.acquireConnection + 348(SQLiteConnectionPool.java:348)
       at android.database.sqlite.SQLiteSession.acquireConnection + 894(SQLiteSession.java:894)
       at android.database.sqlite.SQLiteSession.prepare + 586(SQLiteSession.java:586)
       at android.database.sqlite.SQLiteProgram.<init> + 58(SQLiteProgram.java:58)
       at android.database.sqlite.SQLiteStatement.<init> + 31(SQLiteStatement.java:31)
       at android.database.sqlite.SQLiteDatabase.insertWithOnConflict + 1467(SQLiteDatabase.java:1467)
       at android.database.sqlite.SQLiteDatabase.replaceOrThrow + 1411(SQLiteDatabase.java:1411)
       at com.google.android.exoplayer2.upstream.cache.CacheFileMetadataIndex.set + 166(CacheFileMetadataIndex.java:166)
       at com.google.android.exoplayer2.upstream.cache.SimpleCache.startReadWriteNonBlocking + 422(SimpleCache.java:422)
       at com.google.android.exoplayer2.upstream.cache.SimpleCache.startReadWriteNonBlocking + 49(SimpleCache.java:49)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource + 384(CacheDataSource.java:384)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 317(CacheDataSource.java:317)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
```
Crash 2, happened on multiple devices with system version 8,9 and 10.
```
Fatal Exception: java.lang.StackOverflowError: stack size 1042KB
       at libcore.io.Linux.fstat(Linux.java)
       at libcore.io.BlockGuardOs.fstat + 146(BlockGuardOs.java:146)
       at libcore.io.IoBridge.open + 474(IoBridge.java:474)
       at java.io.RandomAccessFile.<init> + 288(RandomAccessFile.java:288)
       at java.io.RandomAccessFile.<init> + 151(RandomAccessFile.java:151)
       at com.google.android.exoplayer2.upstream.FileDataSource.open + 65(FileDataSource.java:65)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource + 455(CacheDataSource.java:455)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 317(CacheDataSource.java:317)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.StatsDataSource.read + 91(StatsDataSource.java:91)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.readFromDataSource + 260(DefaultExtractorInput.java:260)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.read + 61(DefaultExtractorInput.java:61)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.fillBufferWithAtLeastOnePacket + 386(TsExtractor.java:386)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.read + 276(TsExtractor.java:276)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.feedDataToExtractor + 357(HlsMediaChunk.java:357)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.loadMedia + 325(HlsMediaChunk.java:325)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.load + 301(HlsMediaChunk.java:301)
       at com.google.android.exoplayer2.upstream.Loader$LoadTask.run + 381(Loader.java:381)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1167(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 641(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run + 764(Thread.java:764)

#0. Crashed: Loader:HlsSampleStreamWrapper
       at libcore.io.Linux.fstat(Linux.java)
       at libcore.io.BlockGuardOs.fstat + 146(BlockGuardOs.java:146)
       at libcore.io.IoBridge.open + 474(IoBridge.java:474)
       at java.io.RandomAccessFile.<init> + 288(RandomAccessFile.java:288)
       at java.io.RandomAccessFile.<init> + 151(RandomAccessFile.java:151)
       at com.google.android.exoplayer2.upstream.FileDataSource.open + 65(FileDataSource.java:65)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource + 455(CacheDataSource.java:455)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 317(CacheDataSource.java:317)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.StatsDataSource.read + 91(StatsDataSource.java:91)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.readFromDataSource + 260(DefaultExtractorInput.java:260)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.read + 61(DefaultExtractorInput.java:61)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.fillBufferWithAtLeastOnePacket + 386(TsExtractor.java:386)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.read + 276(TsExtractor.java:276)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.feedDataToExtractor + 357(HlsMediaChunk.java:357)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.loadMedia + 325(HlsMediaChunk.java:325)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.load + 301(HlsMediaChunk.java:301)
       at com.google.android.exoplayer2.upstream.Loader$LoadTask.run + 381(Loader.java:381)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1167(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 641(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run + 764(Thread.java:764)
```
Crash 3, happened on multiple devices with system version 5, 6 and 7.
```
Fatal Exception: java.lang.StackOverflowError: stack size 1038KB
       at libcore.io.Posix.fstat(Posix.java)
       at libcore.io.BlockGuardOs.fstat + 134(BlockGuardOs.java:134)
       at libcore.io.IoBridge.open + 444(IoBridge.java:444)
       at java.io.RandomAccessFile.<init> + 247(RandomAccessFile.java:247)
       at java.io.RandomAccessFile.<init> + 128(RandomAccessFile.java:128)
       at com.google.android.exoplayer2.upstream.FileDataSource.open + 65(FileDataSource.java:65)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource + 455(CacheDataSource.java:455)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 317(CacheDataSource.java:317)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.StatsDataSource.read + 91(StatsDataSource.java:91)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.readFromDataSource + 260(DefaultExtractorInput.java:260)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.read + 61(DefaultExtractorInput.java:61)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.fillBufferWithAtLeastOnePacket + 386(TsExtractor.java:386)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.read + 276(TsExtractor.java:276)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.feedDataToExtractor + 357(HlsMediaChunk.java:357)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.loadMedia + 325(HlsMediaChunk.java:325)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.load + 301(HlsMediaChunk.java:301)
       at com.google.android.exoplayer2.upstream.Loader$LoadTask.run + 381(Loader.java:381)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1133(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 607(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run + 761(Thread.java:761)

#0. Crashed: Loader:HlsSampleStreamWrapper
       at libcore.io.Posix.fstat(Posix.java)
       at libcore.io.BlockGuardOs.fstat + 134(BlockGuardOs.java:134)
       at libcore.io.IoBridge.open + 444(IoBridge.java:444)
       at java.io.RandomAccessFile.<init> + 247(RandomAccessFile.java:247)
       at java.io.RandomAccessFile.<init> + 128(RandomAccessFile.java:128)
       at com.google.android.exoplayer2.upstream.FileDataSource.open + 65(FileDataSource.java:65)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource + 455(CacheDataSource.java:455)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 317(CacheDataSource.java:317)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.cache.CacheDataSource.read + 318(CacheDataSource.java:318)
       at com.google.android.exoplayer2.upstream.StatsDataSource.read + 91(StatsDataSource.java:91)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.readFromDataSource + 260(DefaultExtractorInput.java:260)
       at com.google.android.exoplayer2.extractor.DefaultExtractorInput.read + 61(DefaultExtractorInput.java:61)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.fillBufferWithAtLeastOnePacket + 386(TsExtractor.java:386)
       at com.google.android.exoplayer2.extractor.ts.TsExtractor.read + 276(TsExtractor.java:276)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.feedDataToExtractor + 357(HlsMediaChunk.java:357)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.loadMedia + 325(HlsMediaChunk.java:325)
       at com.google.android.exoplayer2.source.hls.HlsMediaChunk.load + 301(HlsMediaChunk.java:301)
       at com.google.android.exoplayer2.upstream.Loader$LoadTask.run + 381(Loader.java:381)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1133(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 607(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run + 761(Thread.java:761)
```

### [REQUIRED] Version of ExoPlayer being used
2.10.1

### [REQUIRED] Device(s) and version(s) of Android being used
The crashes happened on multiple devices and multiple system versions.
I can't repro the crash by myself.

**Comment by tonihei** [comment]
That's probably a duplicate of #6165. The fix will be part of our next release.

**Comment by Hiwensen** [comment]
Thanks for your replay! @tonihei 
This issue probably looks the same as #6165.
Is there an evaluation date for the next release to fix this issue?

**Comment by tonihei** [comment]
We don't usually give fixed dates for releases, but 2.11.0 is likely to come pretty soon (within weeks).
